src/aquariwm.rs

```diff
@@ -2,99 +2,160 @@
 				xcb::Event::X(x::Event::ConfigureRequest(req)) => {
-					trace!("Configuring window ({})", req.window().resource_id());
-					conn.send_request(&x::ConfigureWindow {
+					trace!(window = req.window().resource_id(), "Configuring window");
+					self.conn.send_request(&x::ConfigureWindow {
 						window: req.window(),
 						value_list: &req.values(),
 					});
 
-					conn.flush()?;
+					self.conn.flush()?;
 				}
 				// Allow clients to map their windows, but initialise those windows with AquariWM
 				// afterwards.
 				xcb::Event::X(x::Event::MapRequest(req)) => {
 					let window = req.window();
-					trace!("Mapping window ({})", window.resource_id());
+					trace!(window = window.resource_id(), "Mapping window");
 
-					conn.send_request(&x::MapWindow { window });
-                    // TODO: Check if this is necessary. Maybe events can be registered fine
-                    //       before a window is mapped, and that it is only that they are cleared
-                    //       when it is unmapped?
-                    conn.flush()?;
+					self.conn.send_request(&x::MapWindow { window });
+					// TODO: Check if this is necessary. Maybe events can be registered fine
+					//       before a window is mapped, and that it is only that they are cleared
+					//       when it is unmapped?
+					self.conn.flush()?;
 
-                    debug!("Setting up newly mapped window ({})", window.resource_id());
-					crate::init_window(conn, &window)?;
-					conn.flush()?;
+					debug!(
+						window = window.resource_id(),
+						"Setting up newly mapped window"
+					);
+					crate::init_window(&self.conn, &window)?;
+					self.conn.flush()?;
 				}
-				// TODO: Unimplemented. For window manipulation.
+				// Start window manipulation if no other window manipulation is already happening.
+				// This event is only fired when the Super key (mod mask 4) is pressed.
 				xcb::Event::X(x::Event::ButtonPress(notif)) => {
-					trace!("Processing button press ({})", notif.detail());
+					if self.manipulation.is_none() {
+						let window = notif.event();
+						let cursor_pos = (notif.root_x(), notif.root_y());
+
+						// If the primary mouse button is pressed, and there is no ongoing window
+						// manipulation, start moving the window.
+						if notif.detail() == x::ButtonIndex::N1 as u8 {
+							self.manipulation =
+								Some(WindowManipulation::moving(&self.conn, window, cursor_pos)?);
+						}
+
+						// If the secondary mouse button is pressed, and there is no ongoing
+						// window manipulation, start resizing the window.
+						if notif.detail() == x::ButtonIndex::N3 as u8 {
+							self.manipulation = Some(WindowManipulation::resizing(
+								&self.conn, window, cursor_pos,
+							)?);
+						}
+					}
 				}
-				// TODO: Unimplemented. For window manipulation.
+				// End window manipulation if one is occurring and its button is released. This
+				// is only fired when the Super key (mod mask 4) is pressed.
```

Comment:
That's a good point. The super key should only be relevant for the button press. There should probably be a separate `GrabPointer` request sent when window manipulation starts, on the manipulated window, that will stop the pointer events going to the window. Don't quote me on the details though - I know there is pointer motion, button motion, grabbing pointer, grabbing buttons... but something along those lines.

Grabbing the buttons/pointer makes it so the events aren't sent to the window itself. These need to be set specifically during window manipulation because, as you say, they should still be grabbed if the Super key is released.

Reply:
As manipulation can only be performed on one window at a time as far as I know, this seems to be a good approach. Grabbing buttons while a manipulation is occuring and waiting for their release, although some UX tests would probably be necessary for it to not feel unnatural

Reply:
> As manipulation can only be performed on one window at a time as far as I know, this seems to be a good approach. Grabbing buttons while a manipulation is occuring and waiting for their release, although some UX tests would probably be necessary for it to not feel unnatural

What do you mean exactly (re: unnatural)? I believe that's how pretty much all window managers do it - the buttons are grabbed because the intention is definitely not to interact with the window. I can imagine that another way to do it would be to make releasing Super cancel the window manipulation instead, though, but I think it makes sense that once you're dragging a window, you still want to be manipulating it until the mouse buttons are released. We could look at how the controls generally work in other window managers - might be bikeshedding though, I know `bspwm` works like the first suggestion where Super only matters to start the window manipulation, and its a small issue overall.